packages/react-components/react-table/src/hooks/selectionManager.ts

```diff
@@ -0,0 +1,112 @@
+import { SelectionMode } from './types';
+
+type OnSelectionChangeCallback = (selectedItems: Set<SelectionItemId>) => void;
+
+export interface SelectionManager {
+  toggleSelect(id: SelectionItemId): void;
+  select(id: SelectionItemId): void;
+  deSelect(id: SelectionItemId): void;
+  clearSelection(): void;
+  isSelected(id: SelectionItemId): boolean;
+  toggleSelectAll(itemIds: SelectionItemId[]): void;
+}
+
+export type SelectionItemId = string | number;
+
+export function createSelectionManager(
+  mode: SelectionMode,
+  onSelectionChange: OnSelectionChangeCallback = () => undefined,
+): SelectionManager {
+  const selectedItems = new Set<SelectionItemId>();
+  return mode === 'multiselect'
+    ? createMultipleSelectionManager(selectedItems, onSelectionChange)
+    : createSingleSelectionManager(selectedItems, onSelectionChange);
+}
+
+function createMultipleSelectionManager(
+  set: Set<SelectionItemId>,
+  onSelectionChange: OnSelectionChangeCallback,
+): SelectionManager {
+  const toggleSelectAll = (itemIds: SelectionItemId[]) => {
+    if (set.size === itemIds.length) {
+      set.clear();
+    } else {
```

Comment:
I understand what is happening there, but is it safe? I can have `Set(['id-1', 'id-2'])`, but I can call `toggleSelectAll(['id-i1', 'id-1'])` as `itemIds` is an array.

Reply:
The main issue is that if I store a reference to the original collection in the manager, it will be a dependency in react, which can invalidate quite easily. I also realized that now you can do this

```ts
toggleSelect('nonexistent')
```

and it would still be managed by the selection manager, if the original collection changes then then the `nonexistent` key will forever by in the selection manager 😨 I guess, there will have to be a dependency on the original collection.

you have any thoughts on that ?

Reply:
I was thinking in this direction:

```ts
  const toggleSelectAll = (itemIds: SelectionItemId[]) => {
    const itemsAreSelected = itemIds.every(itemId => set.has(itemId));

    if (itemsAreSelected) {
      set.clear();
    } else {
      itemIds.forEach(itemId => set.add(itemId));
    }

    onSelectionChange(new Set(set));
  };
```

I agree with you on complications. Let's try the proposed approach without connecting a source collection.

Reply:
yeah that will work, it will still be possible to have a disconnect with the actual data source though (especially if it changes dynamically)

But we can see how that works out in reality